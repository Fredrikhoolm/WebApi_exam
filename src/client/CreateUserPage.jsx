import React, { useState } from "react";
import { InputField } from "./InputField";

export function CreateUserPage({ userApi: userApi }) {
    const [firstName, setFirstName] = useState("");
    const [lastName, setLastName] = useState("");
    const [email, setEmail] = useState("");

    async function submit(e) {
        e.preventDefault();
        await userApi.createUser({ firstName, lastName, email });
    }

    return (
        <form onSubmit={submit}>
            <h1>Create new User</h1>
            <InputField label={"First name"} value={firstName} onChangeValue={setFirstName} />
            <InputField label={"Last name"} value={lastName} onChangeValue={setLastName} />
            <InputField label={"Email"} value={email} onChangeValue={setEmail} />

            <button>Submit</button>
        </form>
    );
}
