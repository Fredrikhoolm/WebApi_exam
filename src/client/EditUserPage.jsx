import React, { useState } from "react";
import { useParams } from "react-router";
import { LoadingView } from "./LoadingView";
import { InputField } from "./InputField";
import { useLoading } from "./useLoading";
import { ErrorView } from "./ErrorView";

function EditUserForm({ user, onSubmit }) {
    const [firstName, setFirstName] = useState(user.firstName);
    const [lastName, setLastName] = useState(user.lastName);
    const [email, setEmail] = useState(user.email);

    async function submit(e) {
        onSubmit(e, { firstName, lastName, email });
    }

    return (
        <form onSubmit={submit}>
            <h1>Edit an existing User ({firstName})</h1>
            <InputField label={"firstName"} value={firstName} onChangeValue={setFirstName} />
            <InputField label={"lastName"} value={lastName} onChangeValue={setLastName} />
            <InputField label={"email"} value={email} onChangeValue={setEmail} />

            <button>Submit</button>
        </form>
    );
}

export function EditUserPage({ userApi }) {
    const { id } = useParams();

    const { data: user, loading, error, reload } = useLoading(
        async () => await userApi.getUser(id),
        [id]
    );

    async function handleSubmit(e, { firstName, lastName, email }) {
        e.preventDefault();
        await userApi.updateUser(id, { firstName, lastName, email  });
    }

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !user) {
        return <LoadingView />;
    }

    return <EditUserForm user={user} onSubmit={handleSubmit} />;
}
