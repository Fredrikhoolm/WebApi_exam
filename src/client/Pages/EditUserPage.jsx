import React, { useState } from "react";
import { useParams } from "react-router";
import { LoadingView } from "../Components/LoadingView";
import { InputField } from "../Components/InputField";
import { useLoading } from "../lib/useLoading";
import { ErrorView } from "../Components/ErrorView";

function EditUserForm({ user, onSubmit }) {
    const [firstName, setFirstName] = useState(user.firstName);
    const [lastName, setLastName] = useState(user.lastName);
    const [email, setEmail] = useState(user.email);

    async function submit(e) {
        onSubmit(e, { firstName, lastName, email });
    }

    return (
        <>
            <div id="header-title">
            <h1>Edit an existing User ({firstName})</h1>
            </div>

        <div id="edit-form">
        <form onSubmit={submit}>
            <InputField label={"First name"} value={firstName} onChangeValue={setFirstName} />
            <br/>
            <InputField label={"Last name"} value={lastName} onChangeValue={setLastName} />
            <br/>
            <InputField label={"Email"} value={email} onChangeValue={setEmail} />
            <br/>
            <button>Submit</button>
        </form>
        </div>
        </>
    );
}

export function EditUserPage({ userApi }) {
    const { id } = useParams();

    const { data: user, loading, error, reload } = useLoading(
        async () => await userApi.getUser(id),
        [id]
    );

    async function handleSubmit(e, { firstName, lastName, email }) {
        e.preventDefault();
        await userApi.updateUser(id, { firstName, lastName, email  });
    }

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !user) {
        return <LoadingView />;
    }

    return <EditUserForm user={user} onSubmit={handleSubmit} />;
}
