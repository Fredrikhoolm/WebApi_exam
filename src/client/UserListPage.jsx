import React from "react";
import { LoadingView } from "./LoadingView";
import { Link } from "react-router-dom";
import { useLoading } from "./useLoading";
import { ErrorView } from "./ErrorView";

export function UserListPage({  userApi }) {
    const { data: user, error, loading, reload } = useLoading(
        async () => await userApi.listUser()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !user) {
        return <LoadingView />;
    }

    return (
        <>
            <h1 id="user-header">List user</h1>
            <div id="list-user">
            {user.map(({ id, firstName }) => (

                <li key={id}>
                    <Link to={`/user/${id}/edit`}>{firstName}</Link>
                </li>
            ))}
            </div>
        </>
    );
}
