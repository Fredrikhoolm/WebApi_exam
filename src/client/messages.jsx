import React from "react";
import { useLoading } from "./useLoading";
import { ErrorView } from "./Components/ErrorView";
import { LoadingView } from "./Components/LoadingView";

export function Messages({ messageApi }) {
    const { data: messages, error, loading, reload } = useLoading(
        async () => await messageApi.listMessages()
    );

    if (error) {
        return <ErrorView error={error} reload={reload} />;
    }

    if (loading || !messages) {
        return <LoadingView />;
    }

    return (
        <>
            {messages.map(({ id, username, message }) => (
                <div key={id}>
                    <h3>To user:{id}</h3>
                    <p>
                        User: {username} Message:{message}
                    </p>
                </div>
            ))}
        </>
    );
}