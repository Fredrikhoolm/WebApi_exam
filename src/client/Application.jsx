import { BrowserRouter, Link } from "react-router-dom";
import {Redirect, Route, Switch} from "react-router";
import * as React from "react";
import { ChatPage } from "./Pages/ChatPage";
import {CreateUserPage} from "./Pages/CreateUserPage";
import {fetchJSON, postJSON} from "./lib/http";
import {UserListPage} from "./Pages/UserListPage";
import {EditUserPage} from "./Pages/EditUserPage";
import {ProfilePage} from "./Pages/ProfilePage";
import {useState} from "react";
import {LoginPage} from "./Pages/LoginPage";
import {LoginCallbackPage} from "./Pages/LoginCallbackPage";

export function Application() {

    const [access_token, setAccess_token ] =useState();

    const googleIdentityProvider = {
        discoveryURL: "https://accounts.google.com/.well-known/openid-configuration",
        client_id: "905480636398-omo1hcghgtgmedidmcffvintfk3cvdm1.apps.googleusercontent.com",
    };

   async  function loadProfile(){
        return fetchJSON("/api/profile", {
            headers: {
            ...(access_token ? {Authorization: `Bearer ${access_token}`} : {}),
            },
        });
    }

    const userApi = {
        listUser: async () => await fetchJSON("/api/user"),
        getUser: async (id) => await fetchJSON(`/api/user/${id}`),
        createUser: async ({ firstName, lastName, email }) => {
            return postJSON("/api/user", {
                method: "POST",
                json: { firstName, lastName, email },
            });
        },
        updateUser: async (id, { firstName, lastName, email }) =>
            postJSON(`/api/user/${id}`, {
                method: "PUT",
                json: { firstName, lastName, email },
            }),
    };


    return (

        <div class="first-page">

        <BrowserRouter>
            <header id="home-button">
                <Link to={"/"}> Home </Link>
            </header>
            <Switch>

                <Route path={"/"} exact>

                    <h1 id="header-title">Home page</h1>

                    <div id="list">

                    <ul>

                        <li id="list-4">
                            <Link to={"/login"}>Login</Link>
                        </li>

                        <li id="list-1" >
                            <Link to={"/profile"} >Profile</Link>
                        </li>

                        <li id="list-2">
                            <Link to={"/create"}>Create</Link>
                        </li>

                        <li id="list-3">
                            <Link to={"/user"}>User</Link>
                        </li>

                        <li id="list-5">
                            <Link to={"/chat"}>Chat</Link>
                        </li>
                    </ul>
                 </div>
                </Route>
                <Route path={"/profile"}>
                    {!access_token ? (<Redirect to ={"/"}/>) :
                   (<ProfilePage loadProfile={loadProfile}/>)}
                </Route>
                <Route path={"/create"}>
                    {!access_token ? (<Redirect to ={"/"}/>) :
                        (<CreateUserPage userApi={userApi}/>)}
                </Route>
                <Route exact path={"/user"}>
                    {!access_token ? (<Redirect to ={"/"}/>) :
                        (<UserListPage userApi={userApi} />)}
                </Route>
                <Route path={"/user/:id/edit"}>
                    {!access_token ? (<Redirect to ={"/"}/>) :
                        (<EditUserPage userApi={userApi}/>)}
                </Route>
                <Route path={"/chat"}>
                    {!access_token ? (<Redirect to ={"/"}/>) :
                        (<ChatPage />)}
                </Route>
                <Route exact path={"/login"}>
                    <LoginPage identityProvider={googleIdentityProvider}></LoginPage>
                </Route>
                <Route path={"/login/callback"}>
                    <LoginCallbackPage identityProvider={googleIdentityProvider} onAccessToken={access_token => setAccess_token(access_token)}/>
                </Route>
                <Route>
                    not found
                </Route>
            </Switch>
        </BrowserRouter>

        </div>
    );
}