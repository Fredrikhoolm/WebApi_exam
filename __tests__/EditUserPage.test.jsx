
import * as ReactDOM from "react-dom";
import * as React from "react";
import { MemoryRouter } from "react-router";
import { act, Simulate } from "react-dom/test-utils";
import {EditUserPage} from "../src/client/EditUserPage";




async function renderForTest(child) {
    const container = document.createElement("div");
    await act(async () => {
        await ReactDOM.render(<MemoryRouter>{child}</MemoryRouter>, container);
    });
    return container;
}

describe("edit user page", () => {
    it("can show information about an existing user", async () => {
        const getUser = () => ({
            firstName: "Niklas",
            lastName: "Ellefsen",
            email: "niklas@ellefsen.no",
        });
        const container = await renderForTest(
            <EditUserPage userApi={{ getUser }} />
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("h1").textContent).toEqual(
            "Edit an existing User (Niklas)"
        );
    });

    it("can show loading screen", async () => {
        const getUser = () => new Promise((resolve) => {});
        const container = await renderForTest(
            <EditUserPage userApi={{ getUser }} />
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual("Loading ...");
    });

    it("can show error message", async () => {
        const getUser = () => {
            throw new Error("Failed to load");
        };
        const container = await renderForTest(
            <EditUserPage userApi={{ getUser}} />
        );
        expect(container.innerHTML).toMatchSnapshot();
        expect(container.querySelector("div").textContent).toEqual(
            "An error occurredError: Failed to load"
        );
    });

    it("updates server on submit", async () => {
        const user = {
            firstName: "Niklas",
            lastName: "Ellefsen",
            email: "niklas@ellefsen.no",
        };
        const getUser = () => user;
        const updateUser = jest.fn();
        const container = await renderForTest(
            <EditUserPage userApi={{ getUser, updateUser }} />
        );
        Simulate.change(container.querySelector("input"), {
            target: { value: "Niklas" },
        });
        Simulate.submit(container.querySelector("form"));
        expect(updateUser).toBeCalledWith(undefined, {
            ...user,
            firstName: "Niklas",
        });
    });
});
