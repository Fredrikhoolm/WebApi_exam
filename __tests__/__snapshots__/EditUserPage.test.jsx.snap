// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit user page can show error message 1`] = `"<div><h1>An error occurred</h1><div>Error: Failed to load</div></div>"`;

exports[`edit user page can show information about an existing user 1`] = `"<div id=\\"header-title\\"><h1>Edit an existing User (Niklas)</h1></div><div id=\\"edit-form\\"><form><div><label>First name: <input type=\\"text\\" value=\\"Niklas\\"></label></div><br><div><label>Last name: <input type=\\"text\\" value=\\"Ellefsen\\"></label></div><br><div><label>Email: <input type=\\"text\\" value=\\"niklas@ellefsen.no\\"></label></div><br><button>Submit</button></form></div>"`;

exports[`edit user page can show loading screen 1`] = `"<div>Loading ...</div>"`;
